### PR TITLE
chore(ui-editable): add TS type in `Editable` component

### PR DESCRIPTION
**Pull request by NamNguyenInstructure**
chore(ui-editable): add TS type in `InplaceEdit` component

### CODE REVIEW

File: packages/ui-editable/src/Editable/props.ts [hunk 1]
```diff
@@ -26,66 +26,102 @@ import PropTypes from 'prop-types'
...
+  onClick: () => void
+  onFocus: () => void
+  onBlur: () => void
+  isVisible: boolean
+  buttonRef: (el: Element) => void
+  readOnly?: boolean
+} & Record<string, any>
 
-type EditableProps = EditableOwnProps
+type renderProps = {
```
Comment (ke1k0) [inline packages/ui-editable/src/Editable/props.ts:59]: Since this is exported from the component, we have to prefix it, so please rename it to `EditableRenderProps`.

File: packages/ui-editable/src/InPlaceEdit/index.tsx [hunk 2]
```diff
@@ -125,13 +126,18 @@ class InPlaceEdit extends Component<InPlaceEditProps> {
...
     label,
     ...buttonProps
+  }: {
+    isVisible: boolean
+    readOnly?: boolean
+    label: string
+    elementRef?: (el: Element | null) => void
+    onClick: () => void
+    onFocus: () => void
+    onBlur: () => void
```
Comment (ke1k0) [inline packages/ui-editable/src/InPlaceEdit/index.tsx:140]: The props will be passed to an IconButton component, so you can type it in a way that it accepts `Partial<IconButtonProps>`, and those extra 3 props we use in our own method.

```suggestion
    }: Partial<IconButtonProps> & {
    label: string
    isVisible: boolean
    readOnly?: boolean
```

File: packages/ui-editable/src/InPlaceEdit/index.tsx [hunk 3]
```diff
@@ -150,14 +156,15 @@ class InPlaceEdit extends Component<InPlaceEditProps> {
...
     getViewerProps,
-    // @ts-expect-error ts-migrate(7031) FIXME: Binding element 'getEditorProps' implicitly has an... Remove this comment to see the full error message
     getEditorProps,
-    // @ts-expect-error ts-migrate(7031) FIXME: Binding element 'getEditButtonProps' implicitly ha... Remove this comment to see the full error message
     getEditButtonProps
+  }: {
+    getContainerProps: renderProps['getContainerProps']
+    getViewerProps: renderProps['getViewerProps']
+    getEditorProps: renderProps['getEditorProps']
+    getEditButtonProps: renderProps['getEditButtonProps']
```
Comment (ke1k0) [inline packages/ui-editable/src/InPlaceEdit/index.tsx:167]: You can simplify this:

```tsx
  renderAll = ({
    getContainerProps,
    getViewerProps,
    getEditorProps,
    getEditButtonProps
  }: renderProps) => { ... }
```

File: packages/ui-editable/src/InPlaceEdit/index.tsx [hunk 4]
```diff
@@ -166,12 +173,16 @@ class InPlaceEdit extends Component<InPlaceEditProps> {
...
+    const { ['ref']: omitRef, ...omitRefFromContainerProps } =
+      getContainerProps()
     return (
       <Flex
         display={this.props.inline ? 'inline-flex' : 'flex'}
         direction={flexDir}
         justifyItems={justifyItems}
-        {...getContainerProps()}
+        {...omitRefFromContainerProps}
+        {...(omitRef as unknown as (el: Flex) => void)}
```
Comment (ke1k0) [inline packages/ui-editable/src/InPlaceEdit/index.tsx:185]: I think this TS error shows that the `ref` is indeed not typed correctly in the `getContainerProps`, and it should be fixed instead. I would revert this change and type the `ref` like this:

```tsx
type containerProps = (props?: Record<string, any>) => {
  ref: React.RefCallback<any>
  ...
}
```

File: packages/ui-editable/src/InPlaceEdit/props.ts [hunk 5]
```diff
@@ -26,39 +26,26 @@ import PropTypes from 'prop-types'
...
 
-type InPlaceEditOwnProps = {
-  renderViewer: (...args: any[]) => any
-  renderEditor: (...args: any[]) => any
-  renderEditButton: (...args: any[]) => any
-  mode: 'view' | 'edit'
-  onChangeMode: (...args: any[]) => any
-  value?: any
-  onChange?: (...args: any[]) => any
+type ExtendeRenderEditButton = {
```
Comment (ke1k0) [inline packages/ui-editable/src/InPlaceEdit/props.ts:31]:
```suggestion
type ExtendedRenderEditButton = {
```

File: packages/ui-editable/src/InPlaceEdit/props.ts [hunk 6]
```diff
@@ -26,39 +26,26 @@ import PropTypes from 'prop-types'
...
+type InPlaceEditOwnProps = {
   /**
    * Function to render the view mode component.
    * It is the consumer's responsibility to provide the
    * current value or children.
    *
    * @returns {element} the viewer DOM sub-tree.
    */
-  renderViewer: PropTypes.func.isRequired,
+  renderViewer: () => JSX.Element
```
Comment (ke1k0) [inline packages/ui-editable/src/InPlaceEdit/props.ts:48]: These `JSX.Element`s should be `React.ReactNode`s

File: packages/ui-editable/src/Editable/props.ts [hunk 7]
```diff
@@ -77,16 +77,20 @@ type EditableOwnProps = {
...
    * @param {Boolean} mode
+   *
    * @param {Function} renderProps.getContainerProps - Props to be spread onto the container element
+   *
    * @param {Function} renderProps.getEditorProps - Props to be spread onto the editor element
+   *
    * @param {Function} renderProps.getEditButtonProps - Props to be spread onto the edit button element
    */
-  children?: (props: renderProps) => JSX.Element
+  children?: (props: EditableRenderProps) => JSX.Element
```
Comment (ke1k0) [inline packages/ui-editable/src/Editable/props.ts:89]: `children` has to be the same as `render` 

```tsx
children?: (props: EditableRenderProps) => React.ReactNode
```

File: packages/ui-editable/src/Editable/props.ts [hunk 8]
```diff
@@ -77,16 +77,20 @@ type EditableOwnProps = {
   onChangeMode: (newMode: string) => void
   /**
    * @param {Object} renderProps
+   *
    * @param {Boolean} mode
```
Comment (ke1k0) [inline packages/ui-editable/src/Editable/props.ts:81]: `* @param {Boolean} renderProps.mode`

File: packages/ui-editable/src/InPlaceEdit/index.tsx [hunk 9]
```diff
@@ -38,6 +39,7 @@ import generateStyle from './styles'
 
 import { propTypes, allowedProps } from './props'
 import type { InPlaceEditProps } from './props'
+import type { EditableRenderProps } from '../Editable/props'
```
Comment (ke1k0) [inline packages/ui-editable/src/InPlaceEdit/index.tsx:42]: @tsProps is missing from this component too

**Review by ke1k0 — CHANGES_REQUESTED** [review]

File: packages/ui-editable/src/Editable/props.ts [hunk 10]
```diff
@@ -21,71 +21,111 @@
...
+   *
    * @param {Function} renderProps.getContainerProps - Props to be spread onto the container element
+   *
    * @param {Function} renderProps.getEditorProps - Props to be spread onto the editor element
+   *
    * @param {Function} renderProps.getEditButtonProps - Props to be spread onto the edit button element
    */
```
Comment (matyasf) [inline packages/ui-editable/src/Editable/props.ts:88]: this looks quite ugly in the docs with the new TS props, I'd refactor it to:
```
  /**
   * Function that you can supply that will return the children of this component.
   * It has one parameter has the following fields:
   * - mode: `view` or `edit`, depending on whether the view or the editor should be rendered.
   * - getContainerProps(props) - Props to be spread onto the container element
   * - getEditorProps(props) - Props to be spread onto the editor element
   * - getEditButtonProps(props) - Props to be spread onto the edit button element
   */
```

Comment (matyasf) [reply packages/ui-editable/src/Editable/props.ts:88]: (same for the other fields, please remove any JSDoc annotations (like `@param`)and use Markdown if needed, since they look ugly

**Review by matyasf — CHANGES_REQUESTED** [review]

**Review by matyasf — CHANGES_REQUESTED** [review]
`InPlaceEdit/props.ts` has still lots of JSDoc in the comments (`@props`, `@returns`,..) rewrite these, so it looks nice in the docs app.

Also you can delete `componentWillUnmount() {}` from `Editable/intex.ts`

Othewise it looks good, nice work!

File: packages/ui-editable/src/Editable/props.ts [hunk 11]
```diff
@@ -71,8 +71,10 @@ type EditableOwnProps = {
...
   mode: 'view' | 'edit'
   /**
-   * Called when the component's mode changes
-   * @param {string} new_mode
+   * Called when the component's mode changes.
+   *
+   * Parameter:
+   * - newMode: string
```
Comment (ke1k0) [inline packages/ui-editable/src/Editable/props.ts:77]: I think you can remove the JsDoc of the parameter here, because the TS type shows it well enough, that it has a string `newMode` param:
_____________________________________
image
_____________________________________

File: packages/ui-editable/src/Editable/props.ts [hunk 12]
```diff
@@ -96,7 +98,9 @@ type EditableOwnProps = {
   value?: any
   /**
    * Called when Editable switches from edit to view mode and the value has changed.
-   * @param {any} value
+   *
+   * Parameter:
+   * - value: any
```
Comment (ke1k0) [inline packages/ui-editable/src/Editable/props.ts:103]: Same here, the TS type is enough.

**Comment by ke1k0** [comment]
When you are finished with the PR, please squash the commits into 1 commit in the end, so we can get rid of the lot of `fix PR` commit messages

**Review by ke1k0 — APPROVED** [review]

**Review by matyasf — APPROVED** [review]
looks good!
